storage: add mvcc rangedels to BenchmarkMVCCExportToSST

tbg opened the pull request:
```
go test ./pkg/storage -run - -timeout 40m -count 10 -benchtime 1s -bench 'MVCCExportToSST' 2>&1 | tee bench.txt && ./bench-compare-params.sh bench.txt numRangeKeys {0,1,100}
```

```
==== old: numRangeKeys=0 new: numRangeKeys=1 ====
name                                                                                       old time/op  new time/op   delta
MVCCExportToSST/numKeys=64/numRevisions=1/exportAllRevisions=false/numRangeKeys=X-24       92.5µs ± 4%  145.7µs ± 9%  +57.51%  (p=0.008 n=5+5)
MVCCExportToSST/numKeys=64/numRevisions=1/exportAllRevisions=true/numRangeKeys=X-24        85.5µs ± 2%  146.2µs ±10%  +71.07%  (p=0.008 n=5+5)
MVCCExportToSST/numKeys=64/numRevisions=10/exportAllRevisions=false/numRangeKeys=X-24       212µs ± 2%    289µs ± 4%  +36.17%  (p=0.008 n=5+5)
MVCCExportToSST/numKeys=64/numRevisions=10/exportAllRevisions=true/numRangeKeys=X-24        351µs ± 4%    455µs ± 2%  +29.53%  (p=0.008 n=5+5)
MVCCExportToSST/numKeys=64/numRevisions=100/exportAllRevisions=false/numRangeKeys=X-24      231µs ± 3%    308µs ± 3%  +33.42%  (p=0.008 n=5+5)
MVCCExportToSST/numKeys=64/numRevisions=100/exportAllRevisions=true/numRangeKeys=X-24      1.63ms ± 2%   2.09ms ± 2%  +27.97%  (p=0.008 n=5+5)
MVCCExportToSST/numKeys=512/numRevisions=1/exportAllRevisions=false/numRangeKeys=X-24       421µs ± 3%    550µs ± 2%  +30.61%  (p=0.008 n=5+5)
MVCCExportToSST/numKeys=512/numRevisions=1/exportAllRevisions=true/numRangeKeys=X-24        387µs ± 2%    517µs ± 2%  +33.74%  (p=0.008 n=5+5)
MVCCExportToSST/numKeys=512/numRevisions=10/exportAllRevisions=false/numRangeKeys=X-24     1.33ms ± 1%   1.84ms ± 1%  +37.58%  (p=0.008 n=5+5)
MVCCExportToSST/numKeys=512/numRevisions=10/exportAllRevisions=true/numRangeKeys=X-24      2.39ms ± 1%   3.17ms ± 2%  +32.63%  (p=0.008 n=5+5)
MVCCExportToSST/numKeys=512/numRevisions=100/exportAllRevisions=false/numRangeKeys=X-24    1.46ms ± 1%   1.97ms ± 1%  +34.70%  (p=0.008 n=5+5)
MVCCExportToSST/numKeys=512/numRevisions=100/exportAllRevisions=true/numRangeKeys=X-24     12.6ms ± 1%   15.9ms ± 1%  +26.51%  (p=0.008 n=5+5)
MVCCExportToSST/numKeys=1024/numRevisions=1/exportAllRevisions=false/numRangeKeys=X-24      761µs ± 2%   1016µs ± 1%  +33.43%  (p=0.008 n=5+5)
MVCCExportToSST/numKeys=1024/numRevisions=1/exportAllRevisions=true/numRangeKeys=X-24       705µs ± 2%    948µs ± 1%  +34.44%  (p=0.008 n=5+5)
MVCCExportToSST/numKeys=1024/numRevisions=10/exportAllRevisions=false/numRangeKeys=X-24    2.64ms ± 1%   3.58ms ± 1%  +35.74%  (p=0.008 n=5+5)
MVCCExportToSST/numKeys=1024/numRevisions=10/exportAllRevisions=true/numRangeKeys=X-24     4.70ms ± 0%   6.25ms ± 1%  +32.90%  (p=0.008 n=5+5)
MVCCExportToSST/numKeys=1024/numRevisions=100/exportAllRevisions=false/numRangeKeys=X-24   2.86ms ± 1%   3.84ms ± 1%  +34.11%  (p=0.008 n=5+5)
MVCCExportToSST/numKeys=1024/numRevisions=100/exportAllRevisions=true/numRangeKeys=X-24    25.3ms ± 2%   31.5ms ± 1%  +24.55%  (p=0.008 n=5+5)
MVCCExportToSST/numKeys=8192/numRevisions=1/exportAllRevisions=false/numRangeKeys=X-24     5.70ms ± 1%   7.41ms ± 1%  +29.97%  (p=0.008 n=5+5)
MVCCExportToSST/numKeys=8192/numRevisions=1/exportAllRevisions=true/numRangeKeys=X-24      5.16ms ± 2%   6.86ms ± 2%  +33.11%  (p=0.008 n=5+5)
MVCCExportToSST/numKeys=8192/numRevisions=10/exportAllRevisions=false/numRangeKeys=X-24    21.2ms ± 2%   29.1ms ± 2%  +37.18%  (p=0.008 n=5+5)
MVCCExportToSST/numKeys=8192/numRevisions=10/exportAllRevisions=true/numRangeKeys=X-24     38.6ms ± 1%   50.0ms ± 1%  +29.55%  (p=0.008 n=5+5)
MVCCExportToSST/numKeys=8192/numRevisions=100/exportAllRevisions=false/numRangeKeys=X-24   24.8ms ± 1%   33.4ms ± 3%  +34.73%  (p=0.008 n=5+5)
MVCCExportToSST/numKeys=8192/numRevisions=100/exportAllRevisions=true/numRangeKeys=X-24     210ms ± 3%    264ms ± 2%  +25.93%  (p=0.008 n=5+5)
MVCCExportToSST/numKeys=65536/numRevisions=1/exportAllRevisions=false/numRangeKeys=X-24    45.1ms ± 1%   58.6ms ± 1%  +30.02%  (p=0.008 n=5+5)
MVCCExportToSST/numKeys=65536/numRevisions=1/exportAllRevisions=true/numRangeKeys=X-24     40.9ms ± 1%   54.1ms ± 1%  +32.40%  (p=0.008 n=5+5)
MVCCExportToSST/numKeys=65536/numRevisions=10/exportAllRevisions=false/numRangeKeys=X-24    183ms ± 1%    244ms ± 0%  +33.19%  (p=0.008 n=5+5)
MVCCExportToSST/numKeys=65536/numRevisions=10/exportAllRevisions=true/numRangeKeys=X-24     314ms ± 0%    413ms ± 2%  +31.33%  (p=0.008 n=5+5)
MVCCExportToSST/numKeys=65536/numRevisions=100/exportAllRevisions=false/numRangeKeys=X-24   260ms ±10%    331ms ±11%  +27.50%  (p=0.008 n=5+5)
MVCCExportToSST/numKeys=65536/numRevisions=100/exportAllRevisions=true/numRangeKeys=X-24    1.83s ± 1%    2.27s ± 1%  +24.03%  (p=0.008 n=5+5)
```

```
==== old: numRangeKeys=0 new: numRangeKeys=100 ====
name                                                                                       old time/op  new time/op    delta
MVCCExportToSST/numKeys=64/numRevisions=1/exportAllRevisions=false/numRangeKeys=X-24       92.5µs ± 4%  5081.4µs ± 0%  +5391.36%  (p=0.008 n=5+5)
MVCCExportToSST/numKeys=64/numRevisions=1/exportAllRevisions=true/numRangeKeys=X-24        85.5µs ± 2%  5083.3µs ± 1%  +5848.68%  (p=0.008 n=5+5)
MVCCExportToSST/numKeys=64/numRevisions=10/exportAllRevisions=false/numRangeKeys=X-24       212µs ± 2%    5283µs ± 1%  +2387.43%  (p=0.008 n=5+5)
MVCCExportToSST/numKeys=64/numRevisions=10/exportAllRevisions=true/numRangeKeys=X-24        351µs ± 4%    5385µs ± 1%  +1432.41%  (p=0.008 n=5+5)
MVCCExportToSST/numKeys=64/numRevisions=100/exportAllRevisions=false/numRangeKeys=X-24      231µs ± 3%    5318µs ± 1%  +2200.66%  (p=0.008 n=5+5)
MVCCExportToSST/numKeys=64/numRevisions=100/exportAllRevisions=true/numRangeKeys=X-24      1.63ms ± 2%    7.06ms ± 1%   +332.92%  (p=0.008 n=5+5)
MVCCExportToSST/numKeys=512/numRevisions=1/exportAllRevisions=false/numRangeKeys=X-24       421µs ± 3%   15413µs ± 1%  +3562.62%  (p=0.008 n=5+5)
MVCCExportToSST/numKeys=512/numRevisions=1/exportAllRevisions=true/numRangeKeys=X-24        387µs ± 2%   15127µs ± 2%  +3810.76%  (p=0.008 n=5+5)
MVCCExportToSST/numKeys=512/numRevisions=10/exportAllRevisions=false/numRangeKeys=X-24     1.33ms ± 1%   17.30ms ± 2%  +1196.69%  (p=0.008 n=5+5)
MVCCExportToSST/numKeys=512/numRevisions=10/exportAllRevisions=true/numRangeKeys=X-24      2.39ms ± 1%   18.16ms ± 2%   +659.80%  (p=0.008 n=5+5)
MVCCExportToSST/numKeys=512/numRevisions=100/exportAllRevisions=false/numRangeKeys=X-24    1.46ms ± 1%   17.67ms ± 1%  +1109.77%  (p=0.008 n=5+5)
MVCCExportToSST/numKeys=512/numRevisions=100/exportAllRevisions=true/numRangeKeys=X-24     12.6ms ± 1%    31.3ms ± 1%   +148.91%  (p=0.008 n=5+5)
MVCCExportToSST/numKeys=1024/numRevisions=1/exportAllRevisions=false/numRangeKeys=X-24      761µs ± 2%   18560µs ± 1%  +2338.25%  (p=0.008 n=5+5)
MVCCExportToSST/numKeys=1024/numRevisions=1/exportAllRevisions=true/numRangeKeys=X-24       705µs ± 2%   18309µs ± 0%  +2497.43%  (p=0.008 n=5+5)
MVCCExportToSST/numKeys=1024/numRevisions=10/exportAllRevisions=false/numRangeKeys=X-24    2.64ms ± 1%   23.30ms ± 1%   +783.25%  (p=0.008 n=5+5)
MVCCExportToSST/numKeys=1024/numRevisions=10/exportAllRevisions=true/numRangeKeys=X-24     4.70ms ± 0%   24.54ms ± 1%   +422.09%  (p=0.008 n=5+5)
MVCCExportToSST/numKeys=1024/numRevisions=100/exportAllRevisions=false/numRangeKeys=X-24   2.86ms ± 1%   23.76ms ± 1%   +729.73%  (p=0.008 n=5+5)
MVCCExportToSST/numKeys=1024/numRevisions=100/exportAllRevisions=true/numRangeKeys=X-24    25.3ms ± 2%    51.2ms ± 1%   +102.28%  (p=0.008 n=5+5)
MVCCExportToSST/numKeys=8192/numRevisions=1/exportAllRevisions=false/numRangeKeys=X-24     5.70ms ± 1%   32.34ms ± 3%   +467.60%  (p=0.008 n=5+5)
MVCCExportToSST/numKeys=8192/numRevisions=1/exportAllRevisions=true/numRangeKeys=X-24      5.16ms ± 2%   31.52ms ± 2%   +511.28%  (p=0.008 n=5+5)
MVCCExportToSST/numKeys=8192/numRevisions=10/exportAllRevisions=false/numRangeKeys=X-24    21.2ms ± 2%    76.5ms ± 2%   +260.02%  (p=0.008 n=5+5)
MVCCExportToSST/numKeys=8192/numRevisions=10/exportAllRevisions=true/numRangeKeys=X-24     38.6ms ± 1%    86.6ms ± 1%   +124.44%  (p=0.008 n=5+5)
MVCCExportToSST/numKeys=8192/numRevisions=100/exportAllRevisions=false/numRangeKeys=X-24   24.8ms ± 1%    86.1ms ± 1%   +247.75%  (p=0.008 n=5+5)
MVCCExportToSST/numKeys=8192/numRevisions=100/exportAllRevisions=true/numRangeKeys=X-24     210ms ± 3%     305ms ± 2%    +45.62%  (p=0.008 n=5+5)
MVCCExportToSST/numKeys=65536/numRevisions=1/exportAllRevisions=false/numRangeKeys=X-24    45.1ms ± 1%    77.3ms ± 2%    +71.41%  (p=0.008 n=5+5)
MVCCExportToSST/numKeys=65536/numRevisions=1/exportAllRevisions=true/numRangeKeys=X-24     40.9ms ± 1%    73.0ms ± 1%    +78.62%  (p=0.008 n=5+5)
MVCCExportToSST/numKeys=65536/numRevisions=10/exportAllRevisions=false/numRangeKeys=X-24    183ms ± 1%     411ms ± 1%   +124.65%  (p=0.008 n=5+5)
MVCCExportToSST/numKeys=65536/numRevisions=10/exportAllRevisions=true/numRangeKeys=X-24     314ms ± 0%     520ms ± 1%    +65.46%  (p=0.008 n=5+5)
MVCCExportToSST/numKeys=65536/numRevisions=100/exportAllRevisions=false/numRangeKeys=X-24   260ms ±10%     857ms ± 3%   +229.61%  (p=0.008 n=5+5)
MVCCExportToSST/numKeys=65536/numRevisions=100/exportAllRevisions=true/numRangeKeys=X-24    1.83s ± 1%     2.63s ± 2%    +44.05%  (p=0.008 n=5+5)
```

Closes https://github.com/cockroachdb/cockroach/issues/86140.

Release note: None
Release justification: testing-only changes

cockroach-teamcity commented:
<!-- Reviewable:start -->
This change is [<img src="https://reviewable.io/review_button.png" height="34" align="absmiddle" alt="Reviewable"/>](https://reviewable.io/reviews/cockroachdb/cockroach/86141)
<!-- Reviewable:end -->

tbg committed:
storage: fix accidental no-op in runMVCCExportToSST

The keys were accidentally prefixed by 16 null bytes, and these sort
before `keys.LocalMax`. As a result, the SST export in this benchmark
never returned any kv pairs, rendering the benchmark useless.

Release note: None

tbg committed:
storage: minor refactor in runMVCCExportToSST

We want to slide MVCC range deletions under the kvs in a follow-up
commit, and this helps make that easier.

Release note: None

tbg committed:
storage: add mvcc rangedels to BenchmarkMVCCExportToSST

Also adds a sanity check that we're not emitting the range dels. This
wasn't actually true, and required a small tweak to the start timestamp
so that the rangedels would actually be excluded.

Closes https://github.com/cockroachdb/cockroach/issues/86140.

tbg committed:
storage: don't throw away output in runMVCCExportToSST

That'll allow us to check it (more on that later).

Release note: None

tbg committed:
storage: skip BenchmarkMVCCExportToSST under short

Release note: None

tbg commented:
RFAL. Note the last commit, where I'm asking of what the expected semantics are - I see range dels in the exported SST even though they're out of time bounds. This happens - sometimes - regardless of what `exportAllVersions` is set to.

I'd like to take this opportunity to improve comments (and if there's a bug, fix it of course)

tbg commented:
I think the comment is actually there, I am still parsing it but things are maybe starting to make sense (https://cockroachlabs.slack.com/archives/C02FTAH8JCT/p1660721713799309)

erikgrinaker commented:
```
benchstat numRangeKeys\={0,100}.txt
MVCCExportToSST/numKeys=64/numRevisions=1/exportAllRevisions=false/X-24        100µs ± 2%    5540µs ± 0%  +5433.82%  (p=0.008 n=5+5)
MVCCExportToSST/numKeys=64/numRevisions=1/exportAllRevisions=true/X-24        90.8µs ± 1%  8408.4µs ± 1%  +9163.79%  (p=0.008 n=5+5)
MVCCExportToSST/numKeys=64/numRevisions=10/exportAllRevisions=false/X-24       220µs ± 4%    5464µs ± 0%  +2379.16%  (p=0.008 n=5+5)
MVCCExportToSST/numKeys=64/numRevisions=10/exportAllRevisions=true/X-24        362µs ± 4%    5574µs ± 1%  +1439.81%  (p=0.008 n=5+5)
MVCCExportToSST/numKeys=64/numRevisions=100/exportAllRevisions=false/X-24      238µs ± 6%    5571µs ± 1%  +2241.91%  (p=0.008 n=5+5)
```

Oof.

tbg committed:
storage: touch up comment on MVCCExportToSST

Release note: None

tbg commented:
The new results are in. Unfortunately they're not much better. For example, this one has a 9k% regression

```
MVCCExportToSST/numKeys=64/numRevisions=1/exportAllRevisions=true/numRangeKeys=100
```

Here are some profiles. From top level, can already see lots of time in `rangekey.coalesce`:

<img width="1662" alt="image" src="https://user-images.githubusercontent.com/5076964/185097396-f3df2cd0-2992-4968-a2f5-9a21b6e1c71e.png">

Zooming into the middle icicle, get more of the same basically:

<img width="1660" alt="image" src="https://user-images.githubusercontent.com/5076964/185097505-f6b4a0f9-ea40-42ad-92a6-e7e328bcecc4.png">

I don't understand why the absolute time with 100 rangekeys gets *better* when we increase `numRevisions`, though. Surely we're doing more work when numRevisions=10 compared to numRevisions=1?

```
MVCCExportToSST/numKeys=64/numRevisions=1/exportAllRevisions=true/X-24        87.8µs ± 1%  8064.0µs ± 1%  +9086.81%  (p=0.008 n=5+5)
MVCCExportToSST/numKeys=64/numRevisions=10/exportAllRevisions=true/X-24        347µs ± 1%    5331µs ± 1%  +1437.46%  (p=0.008 n=5+5)
MVCCExportToSST/numKeys=64/numRevisions=100/exportAllRevisions=true/X-24      1.63ms ± 1%    7.00ms ± 1%   +330.29%  (p=0.008 n=5+5)
```

tbg commented:
> Hmm, yeah, I don't have a explanation for that.

Not sure what the problem was but it's not reproducing, here are some saner numbers:

```
name                                                                                     time/op
MVCCExportToSST/numKeys=64/numRevisions=1/exportAllRevisions=true/numRangeKeys=100-24    5.09ms ± 1%
MVCCExportToSST/numKeys=64/numRevisions=10/exportAllRevisions=true/numRangeKeys=100-24   5.40ms ± 1%
MVCCExportToSST/numKeys=64/numRevisions=100/exportAllRevisions=true/numRangeKeys=100-24  7.04ms ± 0%
```

I'm going to re-run all of them and do another replacement in the top post.

tbg commented:
Done, the numbers are still horrible of course. :-)

erikgrinaker commented on pkg/storage/bench_test.go:
```diff
@@ -1681,6 +1748,37 @@ func runMVCCExportToSST(b *testing.B, numKeys int, numRevisions int, exportAllRe
+	it.SeekGE(MakeMVCCMetadataKey(roachpb.LocalMax))
+	require.NoError(b, err)
+	var n int // points
+	var r int // range key stacks
```
nit: This can be off by as much as 2x, because it's incremented both when we step onto and off of a stack. Still works for our purposes here though.

erikgrinaker reviewed (approved):
LGTM, thanks!

It might be worth having a variant that actually exports the range keys too, and/or one that has lots of garbage below range keys. But this is a good start.

tbg commented:
TFTR!

bors r=erikgrinaker
